pkg/reconciler/finalizer/finalizer.go

```diff
@@ -0,0 +1,298 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package finalizer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/applier"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reposync"
+	"kpt.dev/configsync/pkg/rootsync"
+	"kpt.dev/configsync/pkg/status"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// Finalizer handles finalizing a RootSync or RepoSync for the reconciler.
+type Finalizer struct {
+	Destroyer applier.KptDestroyer
+	Client    client.Client
+
+	// StopControllers will be called by the Finalize() method to stop the Parser and Remediator.
+	StopControllers context.CancelFunc
+
+	// ControllersStopped will be closed by the caller when the parser and
+	// remediator have fully stopped. This unblocks Finalize() to destroy
+	// managed resource objects.
+	ControllersStopped <-chan struct{}
+}
+
+// Finalize performs the following actions on the RootSync or RepoSync:
+// - Stop other controllers
+// - Wait for other controllers to stop
+// - Sets the Finalizing condition
+// - Uses the Destroyer to delete managed objects
+// - Removes the Finalizing condition
+// - Removes the Finalizer (unblocking deletion)
+func (f *Finalizer) Finalize(ctx context.Context, obj *unstructured.Unstructured) error {
+	// Stop the parser & remediator
+	klog.Info("Finalizer scheduled: Parser & Remediator stopping")
+	f.StopControllers()
+
+	// Wait for parser & remediator to stop
+	<-f.ControllersStopped
+	klog.Info("Finalizer executing: Parser & Remediator stopped")
+
+	if _, err := f.setFinalizingCondition(ctx, obj); err != nil {
+		return errors.Wrap(err, "setting Finalizing condition")
+	}
+
+	if err := f.deleteManagedObjects(ctx, obj); err != nil {
+		return errors.Wrap(err, "deleting managed objects")
+	}
+	klog.Infof("Deletion of managed objects successful")
+
+	// TODO: optimize by combining these updates into a single update
+	if _, err := f.removeFinalizingCondition(ctx, obj); err != nil {
+		return errors.Wrap(err, "removing Finalizing condition")
+	}
+	if _, err := f.RemoveFinalizer(ctx, obj); err != nil {
+		return errors.Wrap(err, "removing finalizer")
+	}
+	return nil
+}
+
+// AddFinalizer adds the `configsync.gke.io/reconciler` finalizer to the specified
+// object.
+func (f *Finalizer) AddFinalizer(ctx context.Context, obj *unstructured.Unstructured) (bool, error) {
+	updated, err := updateObjectWithRetry(ctx, f.Client, obj, func() error {
+		if !controllerutil.AddFinalizer(obj, metadata.ReconcilerFinalizer) {
```

Comment:
The `updateObjectWithRetry` function effectively does that. It calls this mutate function immediately and returns if it errors with a `NoUpdateError`.

The reason it's here is actually deliberate, because the `updateObjectWithRetry` function will do a GET if the UPDATE returns a conflict error. This mutate func is then run again against the latest object, which hypothetically could have had its finalizer removed by someone else.

So I can't remove the check from the mutate func. I could add another check before the `updateObjectWithRetry`, but it would be only a very minor optimization.

Reply:
I've replaced `updateObjectWithRetry` with `mutate.WithRetry`, which is similar but more robust. I've added unit tests to show it works, and it does still do the mutation early, before entering the mutation loop. Then it only runs it again if there's another GET, which happens after a ResourceVersion conflict.